ouroboros-network/src/Ouroboros/Network/Connections/Concurrent.hs

```diff
@@ -0,0 +1,143 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE RecursiveDo #-}
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE GADTs #-}
+
+module Ouroboros.Network.Connections.Concurrent
+  ( concurrent
+  ) where
+
+import Control.Concurrent.Async
+import Control.Concurrent.MVar
+import Control.Exception (assert, bracket, finally, mask_)
+import Control.Monad (forM_)
+import Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import Network.Socket (Socket, SockAddr)
+
+import Ouroboros.Network.Connections.Types
+
+data Reject (p :: Provenance) = Duplicate
+
+data Accept (p :: Provenance) = Accept
+
+data Connection
+  = Remote (Async ())
+  | Local (Async ())
+  -- | First is remote, second is local
+  | Both (Async ()) (Async ())
+
+-- | Simple concurrent connection manager as a proof of concept.
+-- It maintains at most 2 connections for each SockAddr: one outgoing and
+-- one incoming. Rejects new ones only if there's already one known to it.
```

Comment:
This is not what is needed. We need *one* connection, and we can have at most *one* connection since the local and remote port will be fixed.

Reply:
Done